Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs

```diff
@@ -1014,60 +1014,125 @@ internal void MigrateJson(ref ReadFileJson parsedJson)
+                    if (parts.Length == 0)
+                        continue;
+
+                    var tokens = new List<string>();
+                    for (int i = 0; i < parts.Length; i += 2)
+                        if (!string.IsNullOrEmpty(parts[i]))
+                            tokens.Add(parts[i]);
+
+                    if (tokens.Count == 0)
+                        continue;
+
+                    var parsed = new List<NameAndParameters>(tokens.Count);
+                    foreach (var t in tokens)
+                        parsed.Add(NameAndParameters.Parse(t));
+
+                    var rebuiltTokens = new List<string>(tokens.Count);
+                    var anyProcessorChanged = false;
+
+                    for (int pi = 0; pi < parsed.Count; pi++)
                     {
-                        var actionJson = mapJson.actions[ai];
-                        var raw = actionJson.processors;
-                        if (string.IsNullOrEmpty(raw))
+                        var nap = parsed[pi];
+
+                        var procType = InputSystem.TryGetProcessor(nap.name);
+                        if (procType == null || nap.parameters.Count == 0)
+                        {
+                            rebuiltTokens.Add(tokens[pi]);
                             continue;
+                        }
+
+                        var dict = new Dictionary<string, string>(nap.parameters.Count, System.StringComparer.OrdinalIgnoreCase);
+                        foreach (var p in nap.parameters)
+                            dict[p.name] = p.value.ToString();
+
+                        var changedThisProcessor = false;
 
-                        var list = NameAndParameters.ParseMultiple(raw).ToList();
-                        var rebuilt = new List<string>(list.Count);
-                        foreach (var nap in list)
+                        foreach (var field in procType.GetFields(BindingFlags.Public | BindingFlags.Instance))
                         {
-                            var procType = InputSystem.TryGetProcessor(nap.name);
-                            if (nap.parameters.Count == 0 || procType == null)
-                            {
-                                rebuilt.Add(nap.ToString());
+                            if (!field.FieldType.IsEnum)
+                                continue;
+
+                            if (!dict.TryGetValue(field.Name, out var rawVal))
                                 continue;
-                            }
 
-                            var dict = nap.parameters.ToDictionary(p => p.name, p => p.value.ToString());
-                            var anyChanged = false;
-                            foreach (var field in procType.GetFields(BindingFlags.Public | BindingFlags.Instance).Where(f => f.FieldType.IsEnum))
+                            if (int.TryParse(rawVal, NumberStyles.Integer, CultureInfo.InvariantCulture, out var n))
```

Comment:
I do not understand what the goal is here, could you add inline comments explaining?